Refatora quase tudo

Eduardo-Barreto opened the pull request:
Salve galera! Ja faz um tempo que estamos comentando que o template não refletia tao bem o jeito que faziamos as coisas na equipe

essa pr aqui basicamente junta tudo que havia de ruim e passa o caminhao

Resumidamente: 
- Migra pra C++
- Migra pra Clang format
- Adiciona Linter
- Adiciona Docker
- Adiciona Doxygen
- Migra pro CMake do CubeMX ao inves do obko
- Muda o uC pra o da bluepill
- Corrige outros erros

Eduardo-Barreto committed:
🔧 Update CMakeLists to cpp

Eduardo-Barreto committed:
⬆️ Update cube

Eduardo-Barreto committed:
✨ Migrate mcu utility to cpp

Eduardo-Barreto committed:
📝 Update README

Eduardo-Barreto committed:
📝 Update file header

Eduardo-Barreto committed:
✨ Update tests to CPP

Eduardo-Barreto committed:
✨ Readd C support

Co-authored-by: Bruno1406 <bruno.lossmachado@gmail.com>

Eduardo-Barreto committed:
🔧 Set uncrustify as default formatter

Eduardo-Barreto committed:
🎨 Function definitions inside namespace

Co-authored-by: Bruno1406 <bruno.lossmachado@gmail.com>

Eduardo-Barreto committed:
🎨 Format with uncrustify

Eduardo-Barreto committed:
📝 Fix guard macros name

Co-authored-by: Lucas Haug <39196309+LucasHaug@users.noreply.github.com>

Eduardo-Barreto committed:
📝 Fix guard macros name comment

Co-authored-by: Lucas Haug <39196309+LucasHaug@users.noreply.github.com>

Eduardo-Barreto committed:
🩹 Rename C_Things to PROJECT_Things

Eduardo-Barreto committed:
🔥 Remove mcu led toggle

Eduardo-Barreto committed:
Merge branch 'migrate/cpp' into the-great-refactor

Eduardo-Barreto committed:
🔧 Change Cube toolchain to CMake

Eduardo-Barreto committed:
🔧 Change uncrustify to clang

Eduardo-Barreto committed:
✨ Add linter

Eduardo-Barreto committed:
♻️ Refactor structure to new cmake

Eduardo-Barreto committed:
🔇 Remove debug logs

Eduardo-Barreto committed:
✨ Add test_all target

Eduardo-Barreto committed:
🚨 Fix linter warnings

Eduardo-Barreto committed:
♻️ Update targets

Eduardo-Barreto committed:
⬆️ Update extensions recommendations

Eduardo-Barreto committed:
🚧 Add doxyfile

Eduardo-Barreto committed:
✨ Add gpio and test

Eduardo-Barreto committed:
♻️ Change uc to bluepill

Eduardo-Barreto committed:
✨🚧 Start to add docker

Eduardo-Barreto committed:
👷 Add CI

Eduardo-Barreto committed:
🚨 Initialize config members

Eduardo-Barreto committed:
🚨🩹 Fix linter warnings with linter

:o

Eduardo-Barreto committed:
📝 Update helpme

Eduardo-Barreto committed:
🐛 Update launch.json template

Eduardo-Barreto committed:
📝 Update README

Eduardo-Barreto committed:
♻️ Update vscode settings

Eduardo-Barreto committed:
➖ Remove stm32-cmake submodule

Eduardo-Barreto committed:
🔧 Set cube code as system

Eduardo-Barreto committed:
🐛🔧 Remove system from cube sources

Eduardo-Barreto committed:
♻️ Use pin and ports from cube

Eduardo-Barreto committed:
🚨💩 Ignore lint error

Eduardo-Barreto committed:
🚨 Reorder includes

Eduardo-Barreto committed:
🔧 Cube dont generate main()

Eduardo-Barreto committed:
🔧🐛 Fix cube includes as system

Eduardo-Barreto committed:
🎨 Make format

Eduardo-Barreto committed:
✨ Add blink

Eduardo-Barreto committed:
🔧🐛 Link stm32cubemx

Eduardo-Barreto committed:
♻️🐛 Fix project sources

Eduardo-Barreto committed:
♻️ Use led toggle

Eduardo-Barreto committed:
♻️🔧 Refactor to use project files as a library

Eduardo-Barreto committed:
📝 omg micras mentioned

Eduardo-Barreto committed:
🩹 Remove thunder_ on sources and headers

Eduardo-Barreto committed:
🔧 Remove lib from format headers and sources

Eduardo-Barreto committed:
🐛 Add whole archive flag to project lib

https://github.com/CrustyAuklet/embedded-cmake-example/tree/master

Eduardo-Barreto committed:
📝♻️ Fix README

Eduardo-Barreto committed:
🔥 Remove english section

ops

jairchtq committed:
:bug: Fix STM32 PROGRAMMER

Eduardo-Barreto committed:
🎨 Remove tab on else statement

d-nery commented on .clang-format:
```diff
@@ -0,0 +1,186 @@
+---
+Language:        Cpp
+BasedOnStyle:  LLVM
```
Mano, ctz que boa parte das coisas desse arquivo já são o default do LLVM asuhsauhas se vai dar o basedOn n precisa explicitar tudo aqui, inclusive tem até uns negócios de macro do QT e do boost (??)

Eduardo-Barreto replied:
ebde7d4

d-nery commented on .github/workflows/ci.yaml:
```diff
@@ -0,0 +1,65 @@
+        uses: actions/cache/restore@v4
+        with:
+            path: docker/Dockerfile
+            key: ${{ hashFiles('.github/*', 'cmake/*', '**/*.cpp', '**/*.hpp', '**/*.ioc', '.clang*', 'CMakeLists.txt', 'docker/*') }}
```
Amei a key que é praticamente ** asuhuashuhs

Eduardo-Barreto replied:
eh papskaodksodoa essencialmente td q impacta no build

d-nery replied:
Olhando de novo, precisa disso? Em que casos você _não_ quer rodar essa pipeline? Tá esquisito um if em tudo asuhasuhasu

Eu to achando que vc na vdd quer mexer  no `on:` lá em cima pra nem triggar em qualquer que seja o caso que tentou aqui

d-nery replied:
Inclusived tem um `if: github.event_name != 'pull_request'` sendo que seria só tirar o o push do on tbm asuhaushasuh

Eduardo-Barreto replied:
eu quero rodar isso só quando tiver alguma coisa que impacta o build

se alguém mexer no readme n tem pq eu ver se o código continua compilando e gastar todos os minutos da organização poaskdpaokfaspok

sobre o `on:` e o if q vc falou, td ja foi tentado iuahfauihfsfh mas ele n eh exatamente oq queríamos

se quiser pode dar uma olhada nessa PR aqui https://github.com/Team-Micras/MicrasFirmware/pull/29 nela tem toda a nossa jornada tentando exatamente isso q vc falou e outras coisas mais kkkkkk

d-nery replied:
Tem muito commit aí sauhasuhash isso aqui não funciona??

```yml
on:
  pull_request:
    branches: ['main', 'develop']
    paths-ignore: ['README.md']
```

E mesmo assim, eu tenho quase ctz que com os ifs o workflow ainda precisa ser iniciado pra chegar no if e isso gasta o tempo de qualquer forma, se der certo no on ele nem começa

Eduardo-Barreto replied:
tem muito commit mesmo aifuhasiuh foi a gente testando muitas variações desses ifs na virada de ano

esse específico q vc mandou foi testado também, _se eu nao me engano_ o problema dele eh que ele só roda qnd a PR eh aberta, e não quando tem novos commits

sobre os minutos, concordo que eh melhor filtrar pelo on, mas se não funcionar, eh melhor eu gastar os segundos até o if do que rodar a pipeline inteira poakspoakfsaopk

GabrielCosme replied:
Na vdd esse ai tem 2 problemas, o primeiro é que quando a PR for mergeada na main ou develop, a action não seria executada na branch, então a branch não ficaria com o checkzinho, e o segundo problema é que ele iria ignorar PRs que envolvam somente o readme, mas se a pr envolver vários arquivos, e eu fizer um push no readme, a action executa de qualquer forma.

d-nery commented on .vscode/extensions.json:
```diff
@@ -6,10 +6,16 @@
+        "twxs.cmake",
+        "ms-vscode.cmake-tools",
+        "seatonjiang.gitmoji-vscode",
+        "mhutchie.git-graph",
```
Eu evitaria colocar QOL aqui, o spell checker, gitmoji e git graph são úteis, mas não necessários pro funcionamento

Eduardo-Barreto replied:
ah eu acho q faz sentido como extensões _recomendadas_, afinal elas ajudam a contribuir com o projeto

Eduardo-Barreto replied:
essencialmente a .vscode inteira n eh _necessária pro funcionamento_ ne aishdaisufh pra mim particularmente ela eh completamente inutil

d-nery replied:
Ah, mais ou menos, as outras geram tasks e configurações pra trabalhar em cima, e as outras extensões tem relação direta com o código e as coisas que o template faz, essas três extensões são coisas mais genéricas e não tem muito a ver. Eu posso por exemplo já ser um usuário do graph do git lens, aí quando eu aceitar o popup de extensões recomendadas vou ter outro graph do nada, ou já ter outro spell checkjer instalado etc

Eduardo-Barreto replied:
eu gosto muito de ter essas recomendações pq o trabalho q alguem novo tem pra ter uma experiencia boa eh simplesmente apertar o botao de instalar as recomendadas

agr se o cara ja usa uma outra extensao q faz o papel de uma que ta ai, eh só ele nn instalar a que a gente recomenda asuifhaiuhas

o trabalho de uma pessoa nova pensar em instalar as extensoes eh maior do que um experiente nao instalar elas

acho q eh esse o papel de vc _recomendar_ extensões, aí vc pode _escolher_ se quer ou nao aopskaopkfasopkf

d-nery commented on CMakeLists.txt:
```diff
@@ -2,179 +2,129 @@
-set(CMAKE_TOOLCHAIN_FILE ${CMAKE_CURRENT_SOURCE_DIR}/lib/stm32-cmake/cmake/stm32_gcc.cmake)
-set(CMAKE_C_STANDARD 11)
+set(CMAKE_C_STANDARD 17)
+set(CMAKE_CXX_STANDARD 20)
```
Oxi põe o 23 logo sauihjasuhasuh

PedroDeSanti replied:
kkkkkk faz q nem o hyprland e bota 26

d-nery commented on Doxyfile:
```diff
@@ -0,0 +1,2860 @@
+# Doxyfile 1.9.8
```
Eu acho isso aqui meio horrível também, ele _precisa_ ter todas as configurações nesse arquivo ou só o que não for default? Pq assim fica muito ruim de achar o que realmente é diferete, se for possível eu deixaria só o que não está default e um simples "For all options see: <link da documentação>" ou algo assim

Eduardo-Barreto replied:
eh uma boa, vou dar uma olhada nisso

coloquei meio q _pra ter_ e pra _alguem_ ver isso um dia kkkkkk

Eduardo-Barreto replied:
398e932

d-nery commented on README.md:
```diff
@@ -1,234 +1,268 @@
+
+Template para projetos com microcontroladores STM32 usando STM32CubeMX e CMake
+
+_Baseado em projetos da ThundeRatz e no [Micras](https://github.com/Team-Micras/MicrasFirmware)_
```
Eu sei que faz sentido, mas eu não colocaria essa linha pq isso é um template, as coisas que deveriam se basear nele asuhasuhs colocaria como créditos/agradecimentos ao final

Eduardo-Barreto replied:
8db7e91

d-nery commented on README.md:
```diff
@@ -1,234 +1,268 @@
+<div align="center">
+  <a href="https://www.st.com/en/development-tools/stm32cubemx.html"><img alt="Usa STM32CubeMX" src="https://img.shields.io/badge/usa-stm32cubemx-blue?style=for-the-badge&labelColor=38c1d0&color=45a4b8" height="30"></a>
+  <a href="https://en.wikipedia.org/wiki/Embedded_system"><img alt="Para Sistemas Embarcados" src="https://img.shields.io/badge/para-sistemas_embarcados-blue?style=for-the-badge&labelColor=adec37&color=27a744" height="30"></a>
+  <a href="LICENSE"><img alt="Licença MIT" src="https://img.shields.io/badge/licença-MIT-blue?style=for-the-badge&labelColor=ef4041&color=c1282d" height="30"></a>
```
Acho que você quer um espaço entre eles, assim tá ficando colado parece uma frase gigante

d-nery replied:
Pera, abrindo o arquivo sozinho até que aparece um espaço, mas no diff aqui ta colado, bom checar se fica certo sempre, de qualquer forma uma margem n faz mal

Eduardo-Barreto replied:
po no render fica normalzinho e no arquivo ta um em cada linha

acho q o diff do github q ficou feio só asiuhfaiusfhsa

d-nery replied:
Ah, é html, estar um em cada linha n quer dizer nada sauhasuhasu
Mas é, eu acho que foi só o diff sim que deixou feio

d-nery commented on README.md:
```diff
@@ -1,234 +1,268 @@
+- **cmake/** - Funções customizadas para CMake
+- **config/** - Configurações do projeto
+- **cube/** - Projeto do STM32CubeMX (.ioc e arquivos gerados)
+- **docker/** - Configurações e scripts do Docker
```
Eu mudaria essa pasta pra `.docker` pra seguir o padrão de pastas com configurações dentro e pra ela não ficar no meio das coisas com código

Eduardo-Barreto replied:
a5ed64d

d-nery commented on README.md:
```diff
@@ -1,234 +1,268 @@
+- **cube/** - Projeto do STM32CubeMX (.ioc e arquivos gerados)
+- **docker/** - Configurações e scripts do Docker
+- **include/** - Cabeçalhos
+- **docs/** - Documentação gerada (não versionado)
```
Eu colocaria as duas pastas não versionadas ao final, tipo

```md
Além disso, as seguintes pastas contém arquivos gerados e não são versionadas:
- build/ - ...
- docs/ - ...
```

Assim a lista principal fica igual ao que é visível no projeto

Dito isso, se não quiser fazer isso, pelo menos colocar o docs no lugar certo na ordem alfabética asuhasuhuash

d-nery commented on CMakeLists.txt:
```diff
@@ -2,179 +2,129 @@
 
-# Cube file name without .ioc extension
-project(stm32_project_template C ASM)
+set(CMAKE_PROJECT_NAME stm32_project_template)
```
Essa variável e a de baixo são as que foram feitas pra serem editadas (inclusive as que tem instrução pra editar no README), é melhor colocar elas bem no começo (só depois do cabeçalho e do minimum required, claro)

d-nery commented on cmake/config_validation.cmake:
Esses .cmake vou só fingir que li

d-nery commented on docker-compose.yaml:
Só um FYI, o mais atual é chamar o arquivo só de `compose.yml`

d-nery reviewed (commented):
(eu troquei de teclado, n sei mais digitar, talvez tenha um monte de letras aleatórias no meio das palavras)

Eduardo-Barreto committed:
📝 Add acknowledgments section

Eduardo-Barreto committed:
🔥 Remove preload

Eduardo-Barreto committed:
🚚 Move docker to .docker

Eduardo-Barreto committed:
💚 Fix docker path on workflows

Eduardo-Barreto committed:
💚 Fix erroneous .docker on image names

PedroDeSanti commented on CMakeLists.txt:
```diff
@@ -2,179 +2,129 @@
+# Debug, Release, RelWithDebInfo and MinSizeRel
 # @see: https://cmake.org/cmake/help/latest/variable/CMAKE_BUILD_TYPE.html
 set(CMAKE_BUILD_TYPE Debug)
 
```
agora isso aqui nao ta servindo pra nada. Ele ta sendo sobrescrito por RelWithDebInfo no config validation

PedroDeSanti commented on cmake/config_validation.cmake:
```diff
@@ -3,69 +3,81 @@
 
     file(WRITE ${CMAKE_CURRENT_BINARY_DIR}/.cube
-        "config load ${CMAKE_CURRENT_SOURCE_DIR}/cube/${PROJECT_RELEASE}.ioc\n"
+        "config load ../cube/${PROJECT_RELEASE}.ioc\n"
```
e se eu quiser criar a minha pasta build em /ababuble/build e dar um cmake ../.. ??kkkkkkkkkkkkk

PedroDeSanti replied:
gosto mais com ${CMAKE_CURRENT_SOURCE_DIR}

PedroDeSanti replied:
além do q vc nao precisa necessariamente criar uma pasta de build kkkk

GabrielCosme replied:
Se n me engano o cmake_source_dir tava quebrando no docker ou no wsl, mas seria bacana encontrar uma solução melhor mesmo, na época eu n tinha achado.

PedroDeSanti replied:
hmmmmmm

d-nery replied:
> e se eu quiser criar a minha pasta build em /ababuble/build e dar um cmake ../.. ??kkkkkkkkkkkkk

Só não querer 🙏

PedroDeSanti commented on cmake/targets.cmake:
```diff
@@ -14,54 +14,80 @@ add_custom_target(helpme
 add_custom_target(reset
     COMMAND echo "Reseting device"
-    COMMAND STM32_Programmer_CLI -c port=SWD -rst
+    COMMAND ${PROGRAMMER_CMD} -c port=SWD -rst
```
aproveita e arruma esse reseting pra resetting

PedroDeSanti commented on cmake/targets.cmake:
```diff
@@ -106,12 +131,12 @@ function(targets_generate_vsfiles_target TARGET)
+    add_custom_target(debug${TARGET_SUFFIX}
         COMMAND echo "Configuring VS Code files for ${TARGET}"
         COMMAND cat ${ouput_save_file} > ${LAUNCH_JSON_PATH}
     )
```
```suggestion
    add_custom_target(debug${TARGET_SUFFIX}
        COMMAND echo "Configuring VS Code files for ${TARGET}"
        COMMAND cat ${output_save_file} > ${LAUNCH_JSON_PATH}
    )
```

PedroDeSanti commented on cmake/config_validation.cmake:
```diff
@@ -3,69 +3,81 @@
 file(GLOB_RECURSE CUBE_SOURCES_CHECK "${CMAKE_CURRENT_SOURCE_DIR}/cube/Src/*.c")
 list(LENGTH CUBE_SOURCES_CHECK CUBE_LENGHT)
 if(CUBE_LENGHT EQUAL 0)
     message(STATUS "Cube directory is empty. Generating cube files...")
```
```suggestion
file(GLOB_RECURSE CUBE_SOURCES_CHECK "${CMAKE_CURRENT_SOURCE_DIR}/cube/Src/*.c")
list(LENGTH CUBE_SOURCES_CHECK CUBE_LENGTH)
if(CUBE_LENGTH EQUAL 0)
    message(STATUS "Cube directory is empty. Generating cube files...")
```

PedroDeSanti replied:
aproveita e arruma aqui tbm

PedroDeSanti commented on src/mcu.cpp:
```diff
@@ -1,32 +1,28 @@
-
-#include "gpio.h"
-#include "main.h"
+#include "mcu.hpp"
+// #include <gpio.h>
+// #include "gpio.h"
+#include "../cube/Inc/gpio.h"
```
qq é isso? kkkkk

Eduardo-Barreto replied:
eu testando coisa q o nery falou e esqueci de tirar kkk

PedroDeSanti commented on CMakeLists.txt:
```diff
@@ -2,179 +2,129 @@
 
 ###############################################################################
-## Include directories
+## Input files
```
Vc tirou a parte do cmake que deixa espaço pronto pra usar lib no projeto. Se pa é bao deixar algo assim

```cmake
file(GLOB_RECURSE PROJECT_LIBS_SOURCES CONFIGURE_DEPENDS "libs/*.c*")

...

# coloca isso aqui pra nao pegar os warnings da lib
set_source_files_properties(
    ${PROJECT_LIBS_SOURCES}
    PROPERTIES
    COMPILE_FLAGS "-w"
)

...

# dai da pra colocar aqui junto
add_library(${PROJECT_NAME}_static STATIC
    ${PROJECT_SOURCES}
    ${PROJECT_LIBS_SOURCES}
)

...

target_include_directories(${PROJECT_NAME}_static PUBLIC
    include
    config
   #include das libs
)
```

Eduardo-Barreto replied:
nossa mano eu acho que eu nao gosto desse approach

eu prefiro usar os

```cmake
add_subdirectory(lib/thunder_proxy)

...

target_link_libraries(${TEST_NAME} PUBLIC
    thunder::proxy
)
```

e aí cada uma ter seu cmakelists

PedroDeSanti commented on README.md:
```diff
@@ -1,234 +1,278 @@
-  > É necessário que a pasta `bin` dessa instalação esteja no `PATH`
+<!-- markdownlint-disable -->
+<div align="center">
+
+# Template para Projetos STM32
+
+Template para projetos com microcontroladores STM32 usando STM32CubeMX e CMake
+
```
Mano nao se sei se é relevante colocar, mas nao tem em nenhum lugar escrito que esse é template usa c++

Eduardo-Barreto replied:
acho q eh bom

Eduardo-Barreto replied:
botei a badge

PedroDeSanti commented on include/gpio.hpp:
```diff
@@ -0,0 +1,75 @@
+/**
+ * @file gpio.hpp
+ *
+ * @brief HAL GPIO class header
+ *
+ * @date 09/2024
+ */
```
```suggestion
/**
 * @file gpio.hpp
 *
 * @brief HAL GPIO class header
 */
```

Tira a data pra ficar no padrao.

PedroDeSanti replied:
ou deixa tudo com data

PedroDeSanti replied:
mas se pa por ser um template faz mais sentido sem data?

GabrielCosme replied:
Particularmente acho que data nunca faz sentido.

VanderSant replied:
ah, eu particulamente acho que não faz sentido botar a data tbm não, pq é só usar o gitlens para saber quem foi o último a editar cada linha e quando. Acho que vale a pena botar a contato da comp da thunder, para facilitar alguém de fora do ciclo thunder/usp tirar alguma duvida:

```suggestion
/**
 * @brief HAL GPIO class header
 *
 * @author Thunderatz Development Team <comp@thunderatz.org>
 */
```

Historicamente na equipe eu sempre senti que a gente sempre usou uns headers que nao tem mt utilidade, tipo `@file gpio.hpp`.  Por mim, eh soh ler o nome do arquivo que vc estah mexendo tlgd haha

Alem disso, @brief  eh meio generico mt vezes, se for para falar algo mt obvio, acho q nem precisar colocar ele. Eu pessoalmente gosto da como os [arquivos do BehaviorTree.CPP](https://github.com/BehaviorTree/BehaviorTree.CPP/blob/master/include/behaviortree_cpp/controls/reactive_fallback.h) sao documentados.

Eduardo-Barreto replied:
se n me engano o doxygen quebra qnd nao tem o `@file`, mas concordo com o resto. Gosto do author thunderatz

PedroDeSanti commented on include/mcu.hpp:
```diff
@@ -0,0 +1,40 @@
+ *****************************************/
+
+extern "C" {
+/**
+ * @brief Initializes System Clock.
+ * @note  Defined by cube.
+ */
+void SystemClock_Config(void);
+}
+
+namespace hal {
+class mcu {
+public:
+    /**
+     * @brief Initializes MCU and some peripherals.
+     */
+    static void init();
+
+    /**
+     * @brief Put the MCU to sleep.
+     *
+     * @param ms  Sleep time in milliseconds
+     */
+    static void sleep(uint32_t ms);
+};
+};  // namespace hal
```
```suggestion
/**
 * @brief Initializes System Clock
 * @note  Defined by cube
 */
void SystemClock_Config(void);
}

namespace hal {
/**
 * @brief Microcontroller unit class
 */
class mcu {
public:
    /**
     * @brief Initializes MCU and some peripherals
     */
    static void init();

    /**
     * @brief Put the MCU to sleep
     *
     * @param ms  Sleep time in milliseconds
     */
    static void sleep(uint32_t ms);
};
};  // namespace hal
```

só pra dedixar no padrao kk

PedroDeSanti replied:
(tirei os pontos finais dos comentarios kkkkk)

PedroDeSanti replied:
e botei um comentario na classe

GabrielCosme replied:
Sou fã de ponto final em tds comentários.

PedroDeSanti replied:
impossivel KKKKKKK

PedroDeSanti replied:
> Sou fã de ponto final em tds comentários.

https://github.com/Team-Micras/MicrasFirmware/commit/b8c93fe4c5c6850fe46feb5b1a042cba827ca939

PedroDeSanti replied:
a hipocrisia

PedroDeSanti replied:
impossivel kkkkkkkkkk

PedroDeSanti replied:
> Sou fã de ponto final em tds comentários.

https://github.com/Team-Micras/MicrasFirmware/commit/b8c93fe4c5c6850fe46feb5b1a042cba827ca939

PedroDeSanti replied:
a hipocrisia

GabrielCosme replied:
https://github.com/Team-Micras/MicrasFirmware/pull/28

Sou um homem mudado.

PedroDeSanti replied:
kkkkkkkkkkk

PedroDeSanti commented on src/gpio.cpp:
```diff
@@ -0,0 +1,31 @@
+    return HAL_GPIO_ReadPin(this->port, this->pin) == this->active_state;
+}
+
+void Gpio::write(bool state) {
+    GPIO_PinState pin_state = this->active_state;
+
+    if (!state) {
+        pin_state = (this->active_state == GPIO_PIN_SET) ? GPIO_PIN_RESET : GPIO_PIN_SET;
+    }
+
+    HAL_GPIO_WritePin(this->port, this->pin, pin_state);
+}
```
```suggestion
void Gpio::write(bool state) {
    GPIO_PinState pin_state = state ? this->active_state : static_cast<GPIO_PinState>(!this->active_state);
    HAL_GPIO_WritePin(this->port, this->pin, pin_state);
}
```

Assim nao fica mais bonitinho? kkkk

Eduardo-Barreto replied:
assim da `Implicit conversion 'GPIO_PinState' -> 'bool'`, pq o `this->active_state` eh um `GPIO_PinState` asuihfaiufhas

eu nao gosto do jeito que ta tambem, mas passei um tempinho brigando com o linter

talvez uma abordagem boa seja apostar na legibilidade e meter um nolint mesmo

PedroDeSanti replied:
da pra usar umas funções privadas pra deixar mais legivel. tipo algo assim:

```cpp
void Gpio::write(bool state) {
    GPIO_PinState pin_state = get_pin_state_from_bool(state);
    HAL_GPIO_WritePin(this->port, this->pin, pin_state);
}

GPIO_PinState Gpio::get_pin_state_from_bool(bool state) const {
    return (state == is_active_high()) ? GPIO_PIN_SET : GPIO_PIN_RESET;
}

bool Gpio::is_active_high() const {
    return this->active_state == GPIO_PIN_SET;
}
```

sei la kkkkk

GabrielCosme replied:
Pq q só n faz uma conversão explícita com static_cast?

PedroDeSanti replied:
tbm da

PedroDeSanti replied:
`static_cast<GPIO_PinState>(!static_cast<bool>(this->active_state))`

nao é acoisa mais linda do mundo kkkk

d-nery replied:
Uai, se ele tá reclamando de implicit conversioin aí é só fazer uma explicit conversion asuhasuhasuhsa

Eduardo-Barreto replied:
> Uai, se ele tá reclamando de implicit conversioin aí é só fazer uma explicit conversion asuhasuhasuhsa

sim po, mas fica feio aaoskfpasfop

d-nery replied:
static_cast é pros feios, os bonitos fazem `bool(this->active_state)` 😘

PedroDeSanti commented on src/gpio.cpp:
```diff
@@ -0,0 +1,31 @@
+ * @brief HAL GPIO class source
+ *
+ * @date 09/2024
+ */
```
```suggestion
/**
 * @file gpio.cpp
 *
 * @brief HAL GPIO class source
 */
```

tira aqui tbm

PedroDeSanti reviewed (commented):
Po, ficou show de bola :) o cmake parece estar check só comentei umas coisinhas. Ah aproveita e atualiza a data do LICENSE

PedroDeSanti commented on README.md:
````diff
@@ -1,234 +1,278 @@
+Para obter uma lista completa de comandos, use
 
-### Compilação
+```bash
+make help
+```
````
```suggestion
```bash
make helpme
```
```

Nao sei se foi proposital, mas make help printa só uma lista de todos os targets, o make helpme printa o textinho bonitinho que ta no cmake/templates/helpme.in

PedroDeSanti commented on README.md:
````diff
@@ -1,234 +1,278 @@
 
-### Compilação
+```bash
+make help
````
```suggestion
make helpme
```

Nao sei se foi proposital, mas make help printa só uma lista de todos os targets, o make helpme printa o textinho bonitinho que ta no cmake/templates/helpme.in

VanderSant replied:
ah, eu lembro dessa discussão em 2023

![image](https://github.com/user-attachments/assets/6d71db35-b36a-4ac1-bdea-1d5479342b05)

d-nery replied:
Nossa, isso n faz sentido nenhum, não é possível que não dê pra desabilitar o default do cmake

PedroDeSanti commented on README.md:
```diff
@@ -1,234 +1,278 @@
-> Os argumentos `C` e `ASM` estão relacionados ao tipo de linguagem que o projeto utiliza (C e Assembly).
-
-Também é necessário alterar as seguintes configurações:
+### Gravando via J-Link
```
```suggestion
### Gravando via [J-Link](https://www.segger.com/downloads/jlink/)
```

Bom colocar aqui o link tbm nao?

VanderSant commented on cube/stm32_project_template.ioc:
```diff
@@ -1,97 +1,47 @@
-NVIC.BusFault_IRQn=true\:0\:0\:false\:false\:false\:true\:false\:false
-NVIC.DebugMonitor_IRQn=true\:0\:0\:false\:false\:false\:true\:false\:false
+Mcu.UserName=STM32F103C8Tx
+MxCube.Version=6.13.0
+MxDb.Version=DB.6.0.130
```
opa, vi aqui que eles jah estao na versao 6.14.0

Eduardo-Barreto replied:
famous last words

VanderSant commented on include/mcu.hpp:
```diff
@@ -0,0 +1,40 @@
+     */
+    static void sleep(uint32_t ms);
+};
+};  // namespace hal
```
```suggestion
};
}  // namespace hal
```

linter achou isso redundante

VanderSant commented on README.md:
```diff
@@ -1,234 +1,278 @@
+Cada teste deve ser um arquivo independente na pasta `test/` com sua própria função `main()`
 
-Com as configurações realizadas corretamente, você deve se direcionar para a pasta `build`. Estando lá, basta rodar o seguinte comando:
+Para compilar um teste específico, use `make meu_teste`. Por exemplo, para compilar o teste `test/test_led.c`:
```
```suggestion
Para compilar um teste específico, use `make meu_teste`. Por exemplo, para compilar o teste `test/test_gpio.cpp`:
```

acho q seria legal botar um teste que realmente existe no repo

VanderSant commented on README.md:
````diff
@@ -1,234 +1,278 @@
 ```bash
-cmake ..
+make test_led
 ```
````
```suggestion
make test_gpio
```
```

VanderSant commented on README.md:
````diff
@@ -1,234 +1,278 @@
 ```
 
-Esse comando é de extrema importância, pois nenhum dos outros comandos de compilação funcionarão sem ele ter sido rodado antes.
+Para gravar um teste específico, use `make flash_meu_teste`:
 
````
isso estah certo? nao seria `make flash meu_teste`?

GabrielCosme replied:
Tem que ser junto msm, não é um parâmetro separado, é um target diferente q ele gera pra cada teste

VanderSant commented on README.md:
````diff
@@ -1,234 +1,278 @@
-> Todos os comandos que envolvam `make` devem ser rodados dentro da pasta `build`, após o comando `cmake ..` ter sido feito.
+```bash
+make flash_test_led
+```
````
```suggestion
```bash
make flash_test_gpio
```
```

VanderSant reviewed (changes requested):
Olás, muito tempo que eu nao reviso uma PR da equipe, mas bora lah! 

Antes de td, muito legal ver o template mudando de novo, desde a minha epoca a gente fala de fazer um versao com docker para validar os MR da equipe. Entao antes de qualquer coisa parabens ai!

Segunda coisa, meu teclado é em inglês e é muito ruim ficar alternando toda hora para acentuar uma palavra ou outra, então perdoa os erros haha 
 
Queria dar uma revisada geral, então eu tentei não ir muito nos detalhes de implementação e focar no macros. Vejo se tentar revisão de novo em outro momento ficando mais nos detalhes

## Execucao Geral

Bom, para começar a revisão, eu decidi seguir o fluxo que uma pessoa totalmente leiga faria para usar o projeto:

- Entrei no meu ubuntu 22.04
- Baixei o cubemx na versão mais recente (6.14.0) e o cmake (3.31.6)
- Segui o readme e deu erro na execução do cmake

![image](https://github.com/user-attachments/assets/87f55863-9f44-4863-acb1-c775f3078d84)

Bom, eu to ligado que isso eh um problema que tem relação com a versão 6.14 do cube, mas alguém que for usar vai dar de cara com esse problema haha Dado isso, acho que seria de bom tom ter um aviso bem destacado no readme sobre não usar essa versão (o que acho que não seria uma solução mt boa a longo prazo 😔) ou corrigir isso antes da release

Depois de mudar a versão para 6.13 o projeto funcionou certinho
## Documentação

Então, a gente sempre misturou mt inglês e ptbr nos projetos, tipo o readme é em prbr, mas o comando de helpme, o codigo, os docs do codigo, ateh mesmo o readme do  stm32 docker são em ingles. 

Dado que esse daqui é um template aberto que não necessariamente soh pessoas da equipe vão usar, acho que agrega valor deixar o readme em inglês direto

## Docker

Bom, eu não entendi mt bem se a ideia do docker era somente compilar uma única vez e usar como verificacao no CICD, mas quando eu executei o service do docker compose e entrei dentro do container, o cmake estava dando um errinho quando eu tentava executar qualquer comando dele:

![image](https://github.com/user-attachments/assets/7bfb8651-dce7-41ac-a2c3-9f2211af6743)


```bash
CMake Error: The current CMakeCache.txt directory /project/build/CMakeCache.txt is different than the directory /home/vanderson/Desktop/codes/thunder/STM32ProjectTemplate/build where CMakeCache.txt was created. This may result in binaries being created in the wrong place. If you are not sure, reedit the CMakeCache.txt
CMake Error: The source "/project/CMakeLists.txt" does not match the source "/home/vanderson/Desktop/codes/thunder/STM32ProjectTemplate/CMakeLists.txt" used to generate cache.  Re-run cmake with a different source directory.
```

GabrielCosme commented:
Respondendo alguma comentários do @VanderSant , esse erro do docker acontece porque o build é feito copiando os arquivos pro container, e o dev utiliza um volume, e ai se vc já rodou um cmake .. na sua máquina, as referências vão estar todas erradas. Mas ai é só apagar o build e dar o cmake .. de novo.

Sobre a versão do cube, diria que é bem complicado fazer funcionar nessa versão nova do cube, teria que refazer grande parte do cmake, e colocar umas gambiarras bem feias, sinceramente não sei nem se é possível sem um script pra alterar o cmake gerado pelo cube. Acredito que a melhor solução é mandar email pra ST, comentar no forum e esperar lançarem uma versão nova.

Eduardo-Barreto committed:
📝 Change make help to helpme

Eduardo-Barreto committed:
🩹 Remove redundant cmake build type

Eduardo-Barreto committed:
🎨 Remove semicolon from namespace ending

Eduardo-Barreto committed:
📝 Change test led to test gpio

Eduardo-Barreto committed:
📝 Update doxystrings on files

Eduardo-Barreto committed:
🔧 Remove date field from doxygen tempalte

Eduardo-Barreto committed:
🐛 Remove test includes

Eduardo-Barreto committed:
🚚 Rename docker compose to compose

Eduardo-Barreto committed:
🎨 Move project configuration up

Eduardo-Barreto committed:
📝 Move ignored folders

Eduardo-Barreto committed:
⬆️ Set cpp version to 26

why not

https://gcc.gnu.org/projects/cxx-status.html#cxx26

d-nery commented:
> não sei nem se é possível sem um script pra alterar o cmake gerado pelo cube. Acredito que a melhor solução é mandar email pra ST, comentar no forum e esperar lançarem uma versão nova.

Ou não usar o cmake do cube rsrsrs

Eduardo-Barreto committed:
🔧 Update build tools

ultimas 3 prs do micras

Eduardo-Barreto committed:
💚 Fix cube cmd on dockerfile

Eduardo-Barreto committed:
🔨 Add devcontainer

Eduardo-Barreto committed:
📝 Add cubemx warning and docker installation

Eduardo-Barreto committed:
📝 Update README

Eduardo-Barreto committed:
📝 Update warning

Eduardo-Barreto committed:
📝 Update windows path instructions

Eduardo-Barreto committed:
🔇 Make docs in batch mode

Eduardo-Barreto committed:
🐛 Fix CUBE_SOURCE_DIR assignment for WSL and Linux environments

https://github.com/Team-Micras/MicrasFirmware/pull/33/commits/de540ec84188e99c16fa70d1ac0faf83f342e1a5

Eduardo-Barreto committed:
🔥 Remove editorconfig

Eduardo-Barreto commented:
> Ou não usar o cmake do cube rsrsrs

![morrer](https://github.com/user-attachments/assets/62f88536-0238-44ad-b0f0-350471d7e3ec)

Eduardo-Barreto committed:
🎨 Fix include orders

Eduardo-Barreto committed:
🎨 Fix line breaking

Eduardo-Barreto committed:
🔧 Optimize clang-format rules

Eduardo-Barreto committed:
🙈 Ignore docs folder

Eduardo-Barreto committed:
🔧 Optimize doxyfile

Eduardo-Barreto committed:
✏️ Fix typo

Eduardo-Barreto committed:
📝 Add english README

Eduardo-Barreto committed:
📄 Update LICENSE

Eduardo-Barreto committed:
♻️ Change githook to clang-format

Eduardo-Barreto committed:
🔥 Remove gitmodules file

Eduardo-Barreto committed:
🩹 Use standard token reference

GabrielCosme commented on .github/workflows/ci.yaml:
```diff
@@ -0,0 +1,65 @@
+name: Build lint and check formatting
+
+on:
+  pull_request:
```
Por padrão o on pull request roda no opened, synchronize e reopened, acho bacana colocar o ready_for_review também, já que ela não roda quando ta em draft.

```suggestion
  pull_request:
    types: [opened, synchronize, reopened, ready_for_review]
```

LucasHaug commented on .github/workflows/ci.yaml:
```diff
@@ -0,0 +1,65 @@
+        if: steps.cache.outputs.cache-hit != 'true'
+        run: docker run --rm project:build /bin/bash /project/.docker/scripts/check_format.sh
+
+      - name: 🔨 Build project
+        if: steps.cache.outputs.cache-hit != 'true'
+        run: docker run --rm project:build /bin/bash /project/.docker/scripts/build.sh
+
+      - name: 🚨 Lint project
+        if: steps.cache.outputs.cache-hit != 'true'
+        run: docker run --rm project:build /bin/bash /project/.docker/scripts/build.sh -DLINTER_MODE=ON
```
Não olhei tão a fundo, mas precisa desses dois estágios? O o build já não é feito junto do linting?

Eduardo-Barreto replied:
verdade, acho q eh melhor juntar nesse caso

a única diferença de resultado eh que quando roda com lint não tem binários de saída, mas como o CI não usa eu acho legal, já eh bom que diminui o tempo do check

Eduardo-Barreto replied:
a gnt tbm perde a clareza da informação se falhou na compilação ou no lint, mas acho q tá suave

LucasHaug replied:
Pra ambos os casos teria que olhar os logs pra ver o que falhou, então acho que não é tanto um problema. E aí economizar tempo na CI sempre ajuda, ainda mais pra projetos grandes aushhshshshsu

d-nery replied:
Eu ainda digo que esse if de cache em todos os steps não faz sentido nenhum e tem que ter uma forma melhor

Eduardo-Barreto replied:
Analisando melhor, acho que vale destacar uma troca importante entre clareza e performance aqui.

Atualmente, temos dois estágios separados: um para o build puro e outro para o build com lint. A proposta de juntar os dois ajuda a economizar tempo no CI, mas tem algumas implicações.

# Comparativo entre os cenários

Build sem linter (estágios separados):

    ⏱️ Build puro: ~23s

    🧹 Lint (com build): ~3min

    🔍 Time-to-fail se o código não compila: 23s

    🔍 Time-to-fail se o lint falha: ~3min 23s

Build com linter somente (estágios unificados):

    🔄 Tudo junto: ~3min

    🔍 Time-to-fail para qualquer falha: sempre até 3min

Ou seja, se o código não compila, o CI gasta 3 minutos para descobrir algo que poderia ser pego em 23 segundos. Isso pode ser relevante principalmente quando estamos em uma iteração rápida ou PRs mais simples.

Além disso, com os estágios separados, a gente ganha um pouco mais de clareza visual no resultado do CI, já que conseguimos ver diretamente se falhou no build ou no lint pelo nome do step (com o emojizinho bonitinho :p). Se unificarmos, isso se perde, aí todo erro vai parecer um erro do "build+lint"

Por outro lado…

É justo dizer que falhas de compilação são menos comuns em branches que sobem PRs, então há um argumento forte pra otimizar pro caso comum (sucesso), e evitar rodar o build puro à toa.

Eu particularmente acho que _tanto faz_ kkkkkkkkkkkkkkkk mas como eu gosto dos emojis tendo a manter como está. De qualquer modo, é uma coisa extremamente simples que pode ser configurada pelo projeto se ele nao gostar do jeito que está

Eduardo-Barreto committed:
💚 Add ready for review trigger on pull request ci

Co-authored-by: Gabriel Cosme Barbosa <62270066+GabrielCosme@users.noreply.github.com>

GabrielCosme commented on cmake/config_validation.cmake:
```diff
@@ -1,71 +1,138 @@
 # Check cube directory for files
 # If it's empty, generate the files
-# It's important to do it before find_package(CMSIS)
 file(GLOB_RECURSE CUBE_SOURCES_CHECK "${CMAKE_CURRENT_SOURCE_DIR}/cube/Src/*.c")
```
```suggestion
file(GLOB_RECURSE CUBE_SOURCES_CHECK "${CMAKE_CURRENT_SOURCE_DIR}/cube/**/*.c")
```

Eduardo-Barreto committed:
🔧🐛 Support for basic and advanced code generation

Co-authored-by: Gabriel Cosme Barbosa <62270066+GabrielCosme@users.noreply.github.com>

GabrielCosme commented on cmake/config_validation.cmake:
```diff
@@ -1,71 +1,138 @@
-else()
-    message(FATAL_ERROR "CUBE_PATH not defined")
+# Check if host system is Linux
+if(NOT CMAKE_HOST_SYSTEM_NAME STREQUAL "Linux")
```
```suggestion
if(NOT CMAKE_HOST_UNIX)
```
Acho que se deixar assim ele suporta MacOS também.

GabrielCosme commented on cmake/config_validation.cmake:
```diff
@@ -1,71 +1,138 @@
-    set(JAVA_EXE $ENV{CUBE_PATH}/jre/bin/java)
-    set(CUBE_JAR $ENV{CUBE_PATH}/STM32CubeMX)
-    set(JLINK_EXE JLinkExe)
+    message(STATUS "Linux detected")
+    set(CUBE_DEFAULT_PATH "/usr/local/STMicroelectronics/STM32Cube/STM32CubeMX")
```
```suggestion
    if (CMAKE_HOST_APPLE)
        message(STATUS "MacOS detected")
        set(CUBE_DEFAULT_PATH "/Applications/STMicroelectronics/STM32CubeMX.app")
    else()
        message(STATUS "Linux detected")
        set(CUBE_DEFAULT_PATH "/usr/local/STMicroelectronics/STM32Cube/STM32CubeMX")
    endif()

```

Tem que pedir pra alguém com Mac testar se só isso é suficiente.

GabrielCosme commented on .github/workflows/ci.yaml:
```diff
@@ -0,0 +1,66 @@
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    if: github.event_name != 'pull_request' || github.event.pull_request.draft == false
+    steps:
```
```suggestion
    if: github.event_name != 'pull_request' || github.event.pull_request.draft == false
    timeout-minutes: 10
    steps:
```

GabrielCosme commented on .github/workflows/ci.yaml:
```diff
@@ -0,0 +1,66 @@
+        uses: docker/setup-buildx-action@v3
+
+      - name: 🐋 Build Docker image
+        if: steps.cache.outputs.cache-hit != 'true'
+        uses: docker/build-push-action@v6
```
```suggestion
        if: steps.cache.outputs.cache-hit != 'true'
        uses: docker/build-push-action@v6
        timeout-minutes: 10
```

GabrielCosme replied:
```suggestion
        if: steps.cache.outputs.cache-hit != 'true'
        timeout-minutes: 10
        uses: docker/build-push-action@v6
```

GabrielCosme commented on .github/workflows/release.yaml:
```diff
@@ -0,0 +1,42 @@
+      - name: 🔧 Set up Docker Buildx
+        uses: docker/setup-buildx-action@v3
+
+      - name: 🐋 Build Docker image
+        uses: docker/build-push-action@v6
```
```suggestion
      - name: 🐋 Build Docker image
        timeout-minutes: 10
        uses: docker/build-push-action@v6
```

PedroDeSanti commented:
<img width="1920" height="1080" alt="image" src="https://github.com/user-attachments/assets/1b0ec5aa-c786-4bd7-9563-74b56ee10b73" />
